src/net/atlas/db.rs

```diff
@@ -0,0 +1,379 @@
+        let tx = self.tx_begin()?;
+
+        for row_text in ATLASDB_SETUP {
+            tx.execute(row_text, NO_PARAMS)
+                .map_err(db_error::SqliteError)?;
+        }
+
+        tx.execute(
+            "INSERT INTO db_version (version) VALUES (?1)",
+            &[&ATLASDB_VERSION],
+        )
+        .map_err(db_error::SqliteError)?;
+
+        tx.commit().map_err(db_error::SqliteError)?;
+
+        Ok(())
+    }
+
+    // Open the burn database at the given path.  Open read-only or read/write.
+    // If opened for read/write and it doesn't exist, instantiate it.
+    pub fn connect(path: &String, readwrite: bool) -> Result<AtlasDB, db_error> {
+        let mut create_flag = false;
+        let open_flags = if fs::metadata(path).is_err() {
+            // need to create
+            if readwrite {
+                create_flag = true;
+                OpenFlags::SQLITE_OPEN_READ_WRITE | OpenFlags::SQLITE_OPEN_CREATE
+            } else {
+                return Err(db_error::NoDBError);
+            }
+        } else {
+            // can just open
+            if readwrite {
+                OpenFlags::SQLITE_OPEN_READ_WRITE
+            } else {
+                OpenFlags::SQLITE_OPEN_READ_ONLY
+            }
+        };
+        let conn =
+            Connection::open_with_flags(path, open_flags).map_err(|e| db_error::SqliteError(e))?;
+
+        let mut db = AtlasDB {
+            conn: conn,
+            readwrite: readwrite,
+        };
+        if create_flag {
+            db.instantiate()?;
+        }
+        Ok(db)
+    }
+
+    // Open a burn database in memory (used for testing)
+    #[cfg(test)]
+    pub fn connect_memory() -> Result<AtlasDB, db_error> {
+        let conn = Connection::open_in_memory().map_err(|e| db_error::SqliteError(e))?;
+
+        let mut db = AtlasDB {
+            conn: conn,
+            readwrite: true,
+        };
+
+        db.instantiate()?;
+        Ok(db)
+    }
+
+    pub fn conn(&self) -> &Connection {
+        &self.conn
+    }
+
+    pub fn tx_begin<'a>(&'a mut self) -> Result<Transaction<'a>, db_error> {
+        if !self.readwrite {
+            return Err(db_error::ReadOnly);
+        }
+
+        let tx = tx_begin_immediate(&mut self.conn)?;
+        Ok(tx)
+    }
+
+    pub fn get_minmax_heights_window_for_page_index(
+        &self,
+        oldest_page_index: u32,
+        newest_page_index: u32,
+    ) -> Result<(u64, u64), db_error> {
+        let qry = "SELECT MIN(block_height) as min, MAX(block_height) as max FROM attachment_instances WHERE page_index >= ?1 AND page_index <= ?2".to_string();
+        let args = [
+            &oldest_page_index as &dyn ToSql,
+            &newest_page_index as &dyn ToSql,
+        ];
+        let mut stmt = self.conn.prepare(&qry)?;
+        let mut rows = stmt.query(&args)?;
+        match rows.next() {
+            Some(Ok(row)) => {
+                let min: i64 = row.get_checked("min")?;
```

Comment:
The reason we don't use `get_checked` elsewhere is because if `min` or `max` _aren't_ i64's, then the database has a serious problem and should probably panic.

Reply:
Addressed with